### ISSUE TITLE
[java] False Alarm of JUnit4TestShouldUseTestAnnotation on Predicates

### ISSUE THREAD

**Issue by havrikov** [pmd/pmd#572]
**Rule Set:** 
`JUnit4TestShouldUseTestAnnotation`

**Description:** 
This rule will falsely report a missing `@Test` annotation on classes implementing the `test` method of the `java.util.Predicate` interface.

**Code Sample demonstrating the issue:**

```java
public final class Whatever<T> implements Predicate<T> {
    @Override
    public boolean test(T t) {
        return false;
    }
}
```

**Running PMD through:** any

**Comment by jsotuyod** [comment]
@havrikov thanks for the report.

Effectively, this migration tools is not checking if the container class is a test (either JUnit 3 or 4).

We should probably rewrite this XPath rule to Java, extending from `AbstractJUnitRule`, and apply the rule only to test classes, regardless of the JUnit version being used.
